rfd/0010-api.md

```diff
@@ -35,89 +34,155 @@ func connectClient() (*auth.Client, error) {
 
-Move relevant lib/client.*Client code to `api/client`.
-Track dependencies, make sure that new `api/client` module has no dependencies
-on the outside modules anywhere in `lib`.
+##### Deprecate the `lib/auth.Client`
 
-**Track dependencies**
+As stated before, all gRPC client related code will be moved out of `teleport/lib/auth/clt.go` and into `api/client`. Backwards compatibility will be maintained by embedding the gRPC `api/client.Client` in the internal `teleport/lib/auth.Client`.
 
-Track dependencies to be limited to:
-  * `gravitational/roundtrip`,  `gravitational/trace`
-  * grpc google package and its dependencies
-  * protobuf generated code
-  * Keep some auth service specific code like `utils.Addr` out of the `api/client`
+Additionally, the remaining HTTP endpoints in the internal client will be converted and added to the gRPC client. To maintain backwards compatibility with old servers, the internal client will be updated to first try the new gRPC endpoint before falling back to the existing HTTP endpoint. These fallback methods will be moved to `teleport/lib/auth/httpfallback.go` and deprecated after one major release cycle.
 
-**Go native user experience**
+There is [issue](https://github.com/gravitational/teleport/issues/6394) used to track remaining endpoints that need to be converted to gRPC. 
 
-Achieve the user experience in Client with a simplified client constructor.
-This should also transparently handle dialing via a proxy.
+Once all necessary endpoints have been added in gRPC and all HTTP fallback logic has been deprecated and removed, the gRPC client can completely replace the existing `lib/auth.Client`. At this point the underlying gRPC server will be the same, so the transition should be relatviley straightforward.
+
+### api module
+
+The api package will be made a sub module of teleport. It will exist in the `teleport/api`, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.
+
+Users will be able to get the latest api version by using `go get github.com/gravitational/teleport/api`, or a specific version with `go get github.com/gravitational/teleport/api@vX.Y.Z`.
+
+#### versioning
+
+Go module versioning must follow semver. Since [Teleport versioning](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md) is a subset of semver, the API can easily follow both version requirements by matching the Teleport version. This means that when `v7.0.1` of Teleport is released, `v7.0.1` of the API will also be released.
+
+Go module versions can automatically be pushed with git tags. Since the api module is a sub module, its version tags will look like `git tag api/vX.Y.Z`. The version can be tagged as part of the release process just like the teleport repository.
```

Comment:
Is this strictly necessary if the API version is always going to be equal to the Teleport version?

Perhaps I'm misunderstanding, but I feel like we'd be better off updating our Teleport `Makefile` logic to just update the version declared in the API's `go.mod` file whenever we do a Teleport release and bundle it under the same tag as Teleport.

Needing to have two separate tags would increase the cognitive load on the release engineer, as the entire release process is still mostly very manual.

Reply:
Unfortunately, yes - https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository

As an example, if you look at at [hashicorp vault](https://github.com/hashicorp/vault), you'll see that `vault/api` and `vault/sdk` are tagged this way, which directly leads to go catching their versions - https://pkg.go.dev/github.com/hashicorp/vault/api?tab=versions

Actually, this looks a bit more complicated though. The main `teleport` module actually isn't following go module rules right now, so it might qualify as a package for the first scenario in the first link above. I'll test this out.

Additionally, we could choose to release the client out of step with Teleport releases. One option would be to release the api as `v1.[major-teleport-version].[api patch/minor version]`. This would allow us to avoid the need for the `vX` suffix, and make it easy to release the api separately from teleport (when needed) while still maintaining the teleport versioning guidelines in a sense (api v1.7.X should work with teleport v7.Y.Z). However solutions such as this move the cognitive load off of release management and onto customers/developing/marketing. But if we want to explore more options like this I can list out some ideas with pros/cons.

Reply:
I tried testing the tags for versioning - https://pkg.go.dev/github.com/Joerger/module-versioning-test.

It looks like submodules have to be tagged separately in our case. Otherwise the [submodule](https://pkg.go.dev/github.com/Joerger/module-versioning-test/submodule/v2) would have a v2.0.0 version available.

Reply:
I see what you mean.

The interesting thing with the Hashicorp example is that they don't release a new SDK/API tag whenever they release a new version of Vault - so while they have lots of `vx.x.x` tags for Vault itself (as Teleport does) they only have a handful of API/SDK tags like `sdk/v1.1.0` and `api/v1.1.0` which makes the tag list much cleaner.

My personal opinion is that we should handle tagging in a similar way to Vault. Although there will be some burden on customers to use the correct API version for their Teleport installation, it should actually make it easier for them overall as they might be able to upgrade through several Teleport versions without needing to change the version of the API they use. We're unlikely to make breaking public API changes unless we absolutely need to, right?

Also, this would definitely be the correct time for us to make this decision as we don't have any public releases of our API at the moment. We could implement everything and release a tag starting off at `v1`, or use the major/minor of Teleport as you suggested.

I think it's worth exploring some different options other than just bumping the API version every time we release Teleport.

Reply:
@webvictim I flushed out the two release options, I think Release Option 1 is actually going to put less burden on release managers, but let me know what you think.